src/socket/udp.rs

```diff
@@ -1,58 +1,74 @@
 use core::cmp::min;
-use managed::Managed;
+use managed::ManagedSlice;
 
 use {Error, Result};
 use wire::{IpProtocol, IpRepr, IpEndpoint, UdpRepr};
 use socket::{Socket, SocketMeta, SocketHandle};
-use storage::{Resettable, RingBuffer};
+use storage::RingBuffer;
 
-/// A buffered UDP packet.
-#[derive(Debug)]
-pub struct PacketBuffer<'a> {
+// Endpoint and size of an UDP packet.
+#[derive(Debug, Clone, Copy, Default)]
+pub struct PacketMetadata {
     endpoint: IpEndpoint,
-    size:     usize,
-    payload:  Managed<'a, [u8]>
+    payload_size: usize,
+    /// Dummy packets can be used to avoid wrap-arounds of packets in the payload buffer
+    dummy: bool,
+}
+
+/// An UDP packet ring buffer.
+#[derive(Debug)]
+pub struct SocketBuffer<'a, 'b> {
+    metadata_buffer: RingBuffer<'a, PacketMetadata>,
+    payload_buffer: RingBuffer<'b, u8>,
 }
 
-impl<'a> PacketBuffer<'a> {
-    /// Create a buffered packet.
-    pub fn new<T>(payload: T) -> PacketBuffer<'a>
-            where T: Into<Managed<'a, [u8]>> {
-        PacketBuffer {
-            endpoint: IpEndpoint::default(),
-            size:     0,
-            payload:  payload.into()
+impl<'a, 'b> SocketBuffer<'a, 'b> {
+    /// Create a new socket buffer with the provided metadata and payload storage.
+    ///
+    /// Metadata storage limits the maximum _number_ of UDP packets in the buffer and payload
+    /// storage limits the maximum _cumulated size_ of UDP packets.
+    pub fn new<MS, PS>(metadata_storage: MS, payload_storage: PS) -> SocketBuffer<'a, 'b>
+        where MS: Into<ManagedSlice<'a, PacketMetadata>>, PS: Into<ManagedSlice<'b, u8>>,
+    {
+        SocketBuffer {
+            metadata_buffer: RingBuffer::new(metadata_storage),
+            payload_buffer: RingBuffer::new(payload_storage),
         }
     }
 
-    fn as_ref<'b>(&'b self) -> &'b [u8] {
-        &self.payload[..self.size]
+    fn is_full(&self) -> bool {
+        self.metadata_buffer.is_full() || self.payload_buffer.is_full()
     }
 
-    fn as_mut<'b>(&'b mut self) -> &'b mut [u8] {
-        &mut self.payload[..self.size]
+    fn is_empty(&self) -> bool {
+        self.metadata_buffer.is_empty() || self.payload_buffer.is_empty()
     }
 
-    fn resize<'b>(&'b mut self, size: usize) -> Result<&'b mut Self> {
-        if self.payload.len() >= size {
-            self.size = size;
-            Ok(self)
+    fn check_capacity(&self, required_size: usize) -> Result<()> {
+        if self.metadata_buffer.is_full() || self.payload_buffer.window() < required_size {
+            Err(Error::Exhausted)
         } else {
-            Err(Error::Truncated)
+                Ok(())
         }
     }
-}
 
-impl<'a> Resettable for PacketBuffer<'a> {
-    fn reset(&mut self) {
-        self.endpoint = Default::default();
-        self.size = 0;
+    fn prepare_for_insert(&mut self, required_size: usize) -> Result<()> {
+        self.check_capacity(required_size)?;
```

Comment:
The control flow here is confusing. I think you can inline one `check_capacity` invocation, remove the second, and rewrite the condition to something like...
```
let window = self.payload_buffer.window();
let contig_window = self.payload_buffer.contiguous_window();
if window < size || window - contig_window < size { return Err(Error::Exhausted) }
```
etc.

Reply:
Also, I suggest calling this something like `enqueue`, moving the non-padding packet enqueueing code in, adding the `enqueue_many` call at the end, and making it return a `Result<&mut [u8]>` with the payload slice, this will reduce duplication below.